src/main/java/de/tum/in/www1/artemis/service/hestia/ExerciseHintService.java

```diff
@@ -0,0 +1,267 @@
+     *
+     * @param template The template exercise containing the hints that should be copied
+     * @param target   The new target exercise, to which all hints should get copied to.
+     * @return A map with the old hint id as a key and the new hint id as a value
+     */
+    public Map<Long, Long> copyExerciseHints(final ProgrammingExercise template, final ProgrammingExercise target) {
+        final Map<Long, Long> hintIdMapping = new HashMap<>();
+        target.setExerciseHints(template.getExerciseHints().stream().map(hint -> {
+            ExerciseHint copiedHint;
+            if (hint instanceof CodeHint) {
+                copiedHint = new CodeHint();
+            }
+            else {
+                copiedHint = new ExerciseHint();
+            }
+
+            copiedHint.setExercise(target);
+            copiedHint.setDescription(hint.getDescription());
+            copiedHint.setContent(hint.getContent());
+            copiedHint.setTitle(hint.getTitle());
+            exerciseHintRepository.save(copiedHint);
+            hintIdMapping.put(hint.getId(), copiedHint.getId());
+            return copiedHint;
+        }).collect(Collectors.toSet()));
+
+        return hintIdMapping;
+    }
+
+    /**
+     * Sets the rating of an exercise hint for a user
+     * The rating is saved in the associated ExerciseHintActivation.
+     *
+     * @param exerciseHint The exercise hint to rate
+     * @param user         The user that submits the rating
+     * @param ratingValue  The value of the rating
+     */
+    public void rateExerciseHint(ExerciseHint exerciseHint, User user, Integer ratingValue) {
+        if (ratingValue < 1 || ratingValue > 5) {
+            throw new BadRequestAlertException("rating has to be between 1 and 5", "exerciseHint", "ratingValue.invalid", false);
+        }
+        var exerciseHintActivation = exerciseHintActivationRepository.findByExerciseHintAndUserElseThrow(exerciseHint.getId(), user.getId());
+        exerciseHintActivation.setRating(ratingValue);
+        exerciseHintActivationRepository.save(exerciseHintActivation);
+    }
+
+    /**
+     * Activates an ExerciseHint for a user.
+     * After activation the user can view the full content of the hint without restrictions.
+     * This action cannot be undone
+     *
+     * @param exerciseHint The exercise hint
+     * @param user         The user
+     * @return true if the hint was activated
+     */
+    public boolean activateHint(ExerciseHint exerciseHint, User user) {
+        // Check if the user has access to the exercise
+        // This is done here to prevent illegal activation of hints in all possible future cases
+        authCheckService.checkHasAtLeastRoleForExerciseElseThrow(Role.STUDENT, exerciseHint.getExercise(), user);
+
+        // Check if the user can activate the hint
+        if (!getAvailableExerciseHints((ProgrammingExercise) exerciseHint.getExercise(), user).contains(exerciseHint)) {
```

Comment:
Can you be sure that the exercise is always a programming exercise?
If yes, would getExercise always return a programming exercise?

Reply:
Yes, it is always a programming exercise. There is no possibility to add exercise hints to other exercise types. I also checked that there are no hints for other types in the production database (thanks to @Mtze).
We will change the getter accordingly.